libqtile/widget/tvheadend.py

```diff
@@ -0,0 +1,327 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import os
+from datetime import datetime
+from pathlib import Path
+
+import requests
+from requests.auth import HTTPBasicAuth
+
+from libqtile import bar, images
+from libqtile.log_utils import logger
+from libqtile.popup import Popup
+from libqtile.widget import base
+
+UPCOMING_RECORDINGS = "/dvr/entry/grid_upcoming"
+
+
+def icon_path():
+    """Get the path to tv icon"""
+    dir_path = Path(__file__).resolve() / ".." / ".." / "resources" / "tvheadend-icons"
+    return str(dir_path.resolve())
+
+
+class TVHTimeOut(Exception):
+    pass
+
+
+class TVHJobServer(object):
+
+    def __init__(self, host=None, auth=None, timeout=5):
+        self.host = host
+        self.auth = auth
+        self.timeout = timeout
+
+    def _send_api_request(self, path, args=None):
+        url = self.host + path
+        try:
+            r = requests.post(url,
+                              data=args,
+                              auth=self.auth,
+                              timeout=self.timeout)
+            return r.json(strict=False)
+        except (requests.exceptions.Timeout, requests.ConnectionError):
+            raise TVHTimeOut()
+
+    def _tidy_prog(self, prog, uuid=None):
+
+        x = prog
+
+        return {"subtitle": x["disp_subtitle"],
+                "title": x["disp_title"],
+                "start_epoch": x["start"],
+                "stop_epoch": x["stop"],
+                "start": datetime.fromtimestamp(x["start"]),
+                "stop": datetime.fromtimestamp(x["stop"]),
+                "filename": x["filename"],
+                "basename": os.path.basename(x["filename"]),
+                "creator": x["creator"],
+                "channelname": x["channelname"],
+                "error": x["errorcode"],
+                "uuid": uuid if uuid else x["uuid"],
+                "duplicate": x.get("duplicate", 0) > 0}
+
+    def get_upcoming(self, hide_duplicates=True):
+        programmes = self._send_api_request(UPCOMING_RECORDINGS)
+        programmes = [self._tidy_prog(x) for x in programmes["entries"]]
+        programmes = sorted(programmes, key=lambda x: x["start_epoch"])
+        if hide_duplicates:
+            programmes = [p for p in programmes if not p["duplicate"]]
+        return programmes
+
+
+class TVHWidget(base._Widget, base.MarginMixin):
+    """
+    A widget to show whether a TVHeadend server is currently recording or not.
+
+    The widget will also show a popup displaying upcoming recordings.
+
+    NB if you use a username and password, these are stored in plain text. You
+    may therefore wish to create an unprivileged user account in TVHeadend that
+    only has access to scheduled recordings data.
```

Comment:
Can you add a note here that this depends on the third party module requests? We've had complaints about this before...